### DISCUSSION TITLE
Feasibility and Stability of Using TinyMPC to Control a Tailsitter VTOL UAV on STM32F405

### DISCUSSION THREAD

**Discussion by jgw365** [TinyMPC/TinyMPC#47] [Q&A]
Hello everyone,

I’m exploring the possibility of using TinyMPC to control a Tailsitter UAV, a type of VTOL aircraft, on an STM32 platform. The Tailsitter is modeled with aerodynamic effects, which introduces nonlinearity and complex dynamics, especially during transitions between hover and forward flight.

I’d like to know if TinyMPC is feasible and stable for this application. Specifically:

Can TinyMPC handle the computational demands of the Tailsitter's aerodynamic model in real-time on an STM32F405?
How robust is TinyMPC when dealing with disturbances or model inaccuracies, particularly during mode transitions?
Any insights or experiences would be greatly appreciated!

**Comment by plancherb1** [comment]
Thanks for the question! TinyMPC is designed specifically to avoid such computationally demanding online dynamics computations. We instead leverage an offline linearization of the dynamics to enable efficient onboard compute. You could possibly try caching a couple of linearizations (e.g., one for hover and one for forward flight) and switch between them (e.g., we have used a couple cached values of the rho regularization parameter in the past). In particular I imagine the hover mode will operate similarly to a quadrotor and so that should work well, and linear MPC originates from the aeronautics field (particularly for fixed wing aircraft and missiles), so I imagine it should work well in the second mode as well. In the transition, that is where I am not as such as to how well it would work. You may need a third or fourth mode to capture the transition dynamics. Alternatively, if the transition is relatively fast, then the fact that we can run TinyMPC at 500hz on an STM32F405 on the Crazyflie suggest that you could run it similarly quickly and potentially overcome any model inaccuracies with fast updates around the transition phase. We didn't subject our demos to wind, however, I have found in the past that fast MPC can overcome moderate disturbances if the model is relatively accurate, and so its worth trying! We are also working on some updates to the solver to make it more robust as well in the future, so hopefully we can get that out soon-ish as that likely would help too!
